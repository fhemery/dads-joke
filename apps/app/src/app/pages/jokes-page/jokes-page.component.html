<div *ngIf="model$ | async; let model">
  <app-jokes-page-header [total]="model.total"></app-jokes-page-header>

  <app-jokes-list *ngIf="model.jokes.length" (favorite)="setFavoriteJoke($event)" [jokes]="model.jokes"></app-jokes-list>
  <div *ngIf="!model.jokes.length" class="no-joke">Oh no!</div>

  <div *ngIf="favoriteJoke" id="favorite-joke">
    <app-jokes-favorite [favorite]="favoriteJoke"></app-jokes-favorite>
  </div>
  dasdasd

</div>
