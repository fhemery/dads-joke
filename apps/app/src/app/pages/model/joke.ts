export interface Joke {
  id: string;
  pun: string;
  isFavorite: boolean;
}
