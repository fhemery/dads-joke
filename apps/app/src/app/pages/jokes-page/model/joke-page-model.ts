import { Joke } from '../../model/joke';

export interface JokePageModel {
  total: number;
  jokes: Joke[];
}
