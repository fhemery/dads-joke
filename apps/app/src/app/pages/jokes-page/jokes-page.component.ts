import { Component, OnDestroy } from '@angular/core';
import { CommonModule } from '@angular/common';
import { JokeDetailComponent } from './joke-detail/joke-detail.component';
import { Joke } from '../model/joke';
import { JokesService } from './jokes.service';
import { Observable, Subject } from 'rxjs';

@Component({
  selector: 'app-jokes-page',
  standalone: true,
  imports: [CommonModule, JokeDetailComponent],
  templateUrl: './jokes-page.component.html',
  styleUrls: ['./jokes-page.component.scss'],
})
export class JokesPageComponent implements OnDestroy {
  jokes$: Observable<Joke[]>;

  private unsubscribe$ = new Subject<void>();

  constructor(private readonly jokesServices: JokesService) {
    this.jokes$ = jokesServices.getJokes();
  }

  ngOnDestroy(): void {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }

  setFavoriteJoke(joke: Joke) {
    console.log('isFavorite called with', joke);
    /*this.jokes = this.jokes.map((j) =>
      j.id === joke.id ? { ...j, isFavorite: true } : j
    );*/
  }

  setUnfavoriteJoke($event: Joke) {
    /*this.jokes = this.jokes.map((j) =>
      j.id === $event.id ? { ...j, isFavorite: false } : j
    );*/
  }
}
