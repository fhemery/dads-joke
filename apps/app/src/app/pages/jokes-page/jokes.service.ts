import { Injectable } from '@angular/core';
import { Joke } from '../model/joke';
import { map, Observable, ReplaySubject, tap } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { JokePageModel } from './model/joke-page-model';

interface ApiJoke {
  id: string;
  joke: string;
}

interface GetJokesApiResponse {
  results: ApiJoke[];
  total_jokes: number;
}

@Injectable({
  providedIn: 'root',
})
export class JokesService {
  private isInitialized = false;
  private jokes$ = new ReplaySubject<JokePageModel>(1);

  constructor(private readonly httpService: HttpClient) {}

  getJokesWithTotal(): Observable<JokePageModel> {
    if (!this.isInitialized) {
      this.httpService
        .get<GetJokesApiResponse>('https://icanhazdadjoke.com/search', {
          headers: { Accept: 'application/json' },
        })
        .pipe(
          map((response) => ({
            total: response.total_jokes,
            jokes: response.results.map((j) => ({
              id: j.id,
              pun: j.joke,
            })),
          })),
          tap((model) => this.jokes$.next(model))
        );
    }
    return this.jokes$;
  }

  getJokes(): Observable<Joke[]> {
    return this.httpService
      .get<GetJokesApiResponse>('https://icanhazdadjoke.com/search', {
        headers: { Accept: 'application/json' },
      })
      .pipe(
        map((response) =>
          response.results.map((j) => ({
            id: j.id,
            pun: j.joke,
          }))
        )
      );
  }
}
