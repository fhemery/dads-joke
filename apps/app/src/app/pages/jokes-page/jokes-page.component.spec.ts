import { ComponentFixture, TestBed } from '@angular/core/testing';

import { JokesPageComponent } from './jokes-page.component';
import { JokesService } from './jokes.service';
import { FavoritesService } from '../favorites-page/favorites.service';
import { JokePageModel } from './model/joke-page-model';
import { Observable, of } from 'rxjs';

describe('JokesPageComponent', () => {
  let component: JokesPageComponent;
  let jokesService: JokesService;
  let favoriteService: FavoritesService;
  let fixture: ComponentFixture<JokesPageComponent>;

  beforeEach(async () => {
    jokesService = {
      getJokesWithTotal(): Observable<JokePageModel> {
        return of({ total: 45, jokes: [] });
      },
    } as JokesService;
    favoriteService = {} as FavoritesService;

    await TestBed.configureTestingModule({
      imports: [JokesPageComponent],
      providers: [
        { provide: JokesService, useValue: jokesService },
        { provide: FavoritesService, useValue: favoriteService },
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(JokesPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should display no-joke card when there are no joke returned', function () {
    const element = fixture.nativeElement.querySelector('.no-joke');
    expect(element).toBeDefined();
    expect(element.textContent).toBe('Oh no!');
  });

  it('should display the favorite joke if it is set', function () {
    component.favoriteJoke = 'Que fait une fraise...';

    fixture.detectChanges();

    const element = fixture.nativeElement.querySelector('#favorite-joke');
    expect(element).toBeTruthy();
  });

  it('should match snapshot', function () {
    component.favoriteJoke = 'Que fait une fraise...';

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
