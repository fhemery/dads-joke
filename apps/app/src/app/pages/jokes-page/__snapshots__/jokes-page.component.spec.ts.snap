// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JokesPageComponent should match snapshot 1`] = `
<app-jokes-page
  __ngContext__="0"
  favoriteJoke={[Function String]}
  favoritesService={[Function Object]}
  jokesServices={[Function Object]}
  model$={[Function Observable]}
>
  <div>
    <app-jokes-page-header>
      <h2>
         Random Jokes (20 / 45)
      </h2>
    </app-jokes-page-header>
    <div
      class="no-joke"
    >
      Oh no!
    </div>
    <div
      id="favorite-joke"
    >
      <app-jokes-favorite>
        <div>
          <h2>
            Your favorite joke
          </h2>
          <div>
            Que fait une fraise...
          </div>
        </div>
      </app-jokes-favorite>
    </div>
  </div>
</app-jokes-page>
`;
